fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/PushRequestedOrderingThroughGroupByRule.java

```diff
@@ -0,0 +1,105 @@
+ */
+public class PushRequestedOrderingThroughGroupByRule extends PlannerRule<GroupByExpression> implements PlannerRule.PreOrderRule  {
+
+    private static final BindingMatcher<ExpressionRef<? extends RelationalExpression>> lowerRefMatcher = ReferenceMatchers.anyRef();
+    @Nonnull
+    private static final BindingMatcher<Quantifier.ForEach> innerQuantifierMatcher = forEachQuantifierOverRef(lowerRefMatcher);
+    @Nonnull
+    private static final BindingMatcher<GroupByExpression> root =
+            groupByExpression(any(innerQuantifierMatcher));
+
+    public PushRequestedOrderingThroughGroupByRule() {
+        super(root, ImmutableSet.of(RequestedOrderingConstraint.REQUESTED_ORDERING));
+    }
+
+    @Override
+    public void onMatch(@Nonnull final PlannerRuleCall call) {
+        final var bindings = call.getBindings();
+        final var groupByExpression = bindings.get(root);
+        final var lowerRef = bindings.get(lowerRefMatcher);
+
+        final var requestedOrderings =
+                call.getPlannerConstraint(RequestedOrderingConstraint.REQUESTED_ORDERING)
+                        .orElse(ImmutableSet.of());
+
+        final var refinedRequestedOrderings = collectCompatibleOrderings(groupByExpression, requestedOrderings);
+
+        if (!refinedRequestedOrderings.isEmpty()) {
+            call.pushConstraint(lowerRef,
+                    RequestedOrderingConstraint.REQUESTED_ORDERING,
+                    refinedRequestedOrderings);
+        }
+    }
+
+    @Nonnull
+    private Set<RequestedOrdering> collectCompatibleOrderings(@Nonnull final GroupByExpression groupByExpression, @Nonnull final Set<RequestedOrdering> requestedOrderings) {
+        final var groupByOrdering = groupByExpression.getOrderingRequirement();
+        // case 1: if no ordering is required, simply specify the group by ordering as a requirement.
+        if (requestedOrderings.isEmpty()) {
+            return Set.of(groupByOrdering);
+        }
+        final var groupByOrderingAsSet = new LinkedHashSet<>(groupByOrdering.getOrderingKeyParts());
+        ImmutableSet.Builder<RequestedOrdering> result = ImmutableSet.builder();
+        for (final var requestedOrdering : requestedOrderings) {
+            if (requestedOrdering.getDistinctness() == RequestedOrdering.Distinctness.PRESERVE_DISTINCTNESS) {
+                // ignore preserve as it might cause e.g. an underlying FUSE operator to be substituted with an access path that is incompatibly-ordered.
+                continue; // ignore
+            }
+            // todo this looks too strict, if requested ordering is more specific (e.g. (a,b,c)) than the requested group by ordering (e.g. (a,b)) should still be accepted.
+            if (!new LinkedHashSet<>(requestedOrdering.getOrderingKeyParts()).equals(groupByOrderingAsSet)) {
+                // case 2: if an incompatible ordering is found, fail.
+                return Set.of(); // fail
```

Comment:
Please use `ImmutableSet` for now as everything in the planner does. We need to switch over at some point I assume but it would have to be done in one fell swoop. The danger of mixing and matching is that both `copyOf()` methods (for `ImmutableSet` in guava, for `Set` in the JDK) have their own bypass logic if the set being passed in is already of type `ImmutableSet` respectively `Set`. If we start mixing and matching we will see a whole bunch of copying going on coming out of nowhere.